appstore/validator.go

```diff
@@ -120,7 +121,20 @@ func (c *Client) Verify(req IAPRequest, result interface{}) error {
 	}
 	defer resp.Body.Close()
 
-	err = json.NewDecoder(resp.Body).Decode(result)
```

Comment:
You could keep this and do a type assertion instead:

```go
r, ok := result.(IAPResponse)
```

Reply:
This line is moved to line 148. The line still exists. I believe the reason that `Verify` takes in an interface{} is to be flexible. Thats why the only time we need an IAPResponse is to check for 21007, but it should not return an error if there is an error (meaning that the error should not affect the outcome of the function)

Reply:
You're now unmarshaling twice even if you don't get a status code 21007.

`json.NewDecoder` is great for when you have a reader, it'll stream it into a buffer for you (instead of you using `ioutil.ReadAll()` to do it yourself).

It's hard to explain in words but this is what I would do:

```golang
	err = json.NewDecoder(resp.Body).Decode(result)
	if err != nil {
		return err
	}

	// Always verify your receipt first with the production URL; proceed to verify with the sandbox URL if you receive
	// a 21007 status code
	//
	// https://developer.apple.com/library/content/technotes/tn2413/_index.html#//apple_ref/doc/uid/DTS40016228-CH1-RECEIPTURL
	r, ok := result.(IAPResponse)
	if ok && r.Status == 21007 {
		resp, err := client.Post(SandboxURL, "application/json; charset=utf-8", b)
		if err != nil {
			return err
		}
		defer resp.Body.Close()

		return json.NewDecoder(resp.Body).Decode(result)
	}

	return nil
```

This won't change the result of the function. If it couldn't assert the type as `IAPResponse`, like before, it won't return an error.